### PR TITLE
Improve Proxmox volume handling

### LINKED ISSUE #8407: Proxmox LXC - Mounts configuration causes repeated creation of volumes
### Summary

Currently, it seems the creation of mount points with the documented syntax is not idempotent but instead creates a new mount disk each time the update task is run.

In the [documentation](https://docs.ansible.com/ansible/latest/collections/community/general/proxmox_module.html)'s example, the syntax is:
`<storage>:<size-in-GB>,mp=<mountpoint>`

### Issue Type

Bug Report

### Component Name

proxmox

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.16.7]
  config file = <my-repo>/ansible.cfg
  configured module search path = ['<my-home>/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = <my-repo>/.venv/lib/python3.12/site-packages/ansible
  ansible collection location = <my-home>/.ansible/collections:/usr/share/ansible/collections
  executable location = <my-repo>/.venv/bin/ansible
  python version = 3.12.3 (main, Apr 23 2024, 09:16:07) [GCC 13.2.1 20240417] (<my-repo>/.venv/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general
Collection        Version
----------------- -------
community.general 8.4.0
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed
CONFIG_FILE() = <my-repo>/ansible.cfg
DEFAULT_ROLES_PATH(<my-repo>/ansible.cfg) = ['<my-repo>/roles']
DEFAULT_TIMEOUT(<my-repo>/ansible.cfg) = 30
EDITOR(env: EDITOR) = vim
PAGER(env: PAGER) = less
```

### OS / Environment

```
proxmoxer==2.0.1
requests==2.32.2
```

### Steps to Reproduce

```yaml (paste below)
- name: Create LXC
  community.general.proxmox:
    api_host: "{{ api_host }}"
    api_user: "{{ api_user }}"
    api_token_id: "{{ api_token_id }}"
    api_token_secret: "{{ api_token_secret }}"
    hostname: "{{ inventory_hostname }}"
    node: my-node
    ostemplate: images:vztmpl/debian-12-standard_12.2-1_amd64.tar.zst
    storage: machines01
    timeout: 30
  register: create_result

- name: Update LXC config
  community.general.proxmox:
    api_host: "{{ api_host }}"
    api_user: "{{ api_user }}"
    api_token_id: "{{ api_token_id }}"
    api_token_secret: "{{ api_token_secret }}"
    node: my-node
    vmid: "{{ create_result.vmid }}"
    cores: 2
    memory: 2048
    swap: 2048
    disk: machines01:vm-{{ create_result.vmid }}-disk-0,size=12G"
    mounts: 
      mp0: machines01:4,mp=/mnt/data
    update: true
```

### Expected Results

The LXC is created and configured with a single mountpoint idempotently, even when the playbook is run again.
More importantly, no extra volumes are created on subsequent runs unless specified.

### Actual Results

The LXC is updated and a new mount volume is created. On a container restart, the previous volume is unmounted and the newly-created on is mounted.
In the `machines01` storage, new volumes `vm-<id>-disk-<n>` are created each time the playbook is run.
In the LXC resources tab, the actually mounted disk is `machines01:vm-<id>-disk-<n>,mp=/mnt/data,size=4G` with a change to `machines01:4,mp=/mnt/data` is queued.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by Lithimlin**
##### SUMMARY

Fixes #8407

Makes the creation of new volumes idempotent and adds new keys with better readability.

##### ISSUE TYPE

- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME

proxmox

##### ADDITIONAL INFORMATION

Here is a sample playbook to test the new features/bug fixes:

```yaml
---
- name: Create LXC
  hosts: localhost
  become: false
  gather_facts: false
  vars_files:
    - proxmox_credentials.yml
  tasks:
    - name: Create LXC
      community.general.proxmox:
        api_host: "{{ api_host }}"
        api_user: "{{ api_user }}"
        api_password: "{{ api_password }}"
        # api_token_id: "{{ api_token_id }}"
        # api_token_secret: "{{ api_token_secret }}"
        node: "{{ node }}"
        hostname: test-node
        ostemplate: local:vztmpl/debian-12-standard_12.2-1_amd64.tar.zst
        disk: local-lvm:8
      register: create_res

    - name: Update LXC config [new - create volumes]
      community.general.proxmox:
        api_host: "{{ api_host }}"
        api_user: "{{ api_user }}"
        api_password: "{{ api_password }}"
        # api_token_id: "{{ api_token_id }}"
        # api_token_secret: "{{ api_token_secret }}"
        node: "{{ node }}"
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk_volume:
          storage: local-lvm
          size: 14
        mount_volumes:
          - id: mp0
            storage: local-lvm
            size: 3
            mountpoint: /mnt/test
          - id: mp1
            host_path: /root/test
            mountpoint: /mnt/host
            options:
              backup: "0"
        update: true

    - name: Update LXC config [new - explicit volumes]
      community.general.proxmox:
        api_host: "{{ api_host }}"
        api_user: "{{ api_user }}"
        api_password: "{{ api_password }}"
        # api_token_id: "{{ api_token_id }}"
        # api_token_secret: "{{ api_token_secret }}"
        node: "{{ node }}"
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk_volume:
          storage: local-lvm
          volume: "vm-{{ create_res.vmid }}-disk-0"
          size: 14
        mount_volumes:
          - id: mp0
            storage: local-lvm
            volume: "vm-{{ create_res.vmid }}-disk-1"
            size: 3
            mountpoint: /mnt/test
          - id: mp1
            host_path: /root/test
            mountpoint: /mnt/host
            options:
              backup: "0"
        update: true

    - name: Update LXC config [old - implicit volumes]
      community.general.proxmox:
        api_host: "{{ api_host }}"
        api_user: "{{ api_user }}"
        api_password: "{{ api_password }}"
        # api_token_id: "{{ api_token_id }}"
        # api_token_secret: "{{ api_token_secret }}"
        node: "{{ node }}"
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk: local-lvm:14
        mounts: '{"mp0":"local-lvm:3,mp=/mnt/test","mp1":"/root/test,mp=/mnt/host,backup=0"}'
        update: true

    - name: Update LXC config [old - explicit volumes]
      community.general.proxmox:
        api_host: "{{ api_host }}"
        api_user: "{{ api_user }}"
        api_password: "{{ api_password }}"
        # api_token_id: "{{ api_token_id }}"
        # api_token_secret: "{{ api_token_secret }}"
        node: "{{ node }}"
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk: "local-lvm:vm-{{ create_res.vmid }}-disk-0,size=14G"
        mounts: '{"mp0":"local-lvm:vm-{{ create_res.vmid }}-disk-1,size=3G,mp=/mnt/test","mp1":"/root/test,mp=/mnt/host,backup=0"}'
        update: true
```

Note: To test bind mounting, we actually need to use the `root@pam` user with its password. Even a token owned by that user will not work due to missing permissions.

Output:
```
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Create LXC] *******************************************************************************************************************************************************************

TASK [Create LXC] *******************************************************************************************************************************************************************
ok: [localhost]

TASK [Update LXC config [new - create volumes]] *************************************************************************************************************************************
changed: [localhost]

TASK [Update LXC config [new - explicit volumes]] ***********************************************************************************************************************************
ok: [localhost]

TASK [Update LXC config [old - implicit volumes]] ***********************************************************************************************************************************
ok: [localhost]

TASK [Update LXC config [old - explicit volumes]] ***********************************************************************************************************************************
ok: [localhost]

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=5    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0  
```

### CODE REVIEW

File: changelogs/fragments/8542-fix-proxmox-volume-handling.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - proxmox - fix idempotency on creation of mount volumes using Proxmox' special ``<storage>:<size>`` syntax.
+minor_changes:
+  - proxmox - add ``disk_volume`` and ``mount_volumes`` keys for better readability.
+  - proxmox - translate the old ``disk`` and ``mounts`` keys to the new handling internally.
```
Comment (felixfontein) [inline changelogs/fragments/8542-fix-proxmox-volume-handling.yml:5]:
```suggestion
  - proxmox - fix idempotency on creation of mount volumes using Proxmox' special ``<storage>:<size>`` syntax (https://github.com/ansible-collections/community.general/issues/8407, https://github.com/ansible-collections/community.general/pull/8542).
minor_changes:
  - proxmox - add ``disk_volume`` and ``mount_volumes`` keys for better readability (https://github.com/ansible-collections/community.general/pull/8542).
  - proxmox - translate the old ``disk`` and ``mounts`` keys to the new handling internally (https://github.com/ansible-collections/community.general/pull/8542).
```

File: plugins/modules/proxmox.py [hunk 2]
```diff
@@ -51,6 +52,41 @@
...
+  disk_volume:
+    description:
+      - Specify a hash/dictionary of the C(rootfs) disk.
+      - See U(https://pve.proxmox.com/wiki/Linux_Container#pct_mount_points) for a full description.
+    type: dict
+    suboptions:
+      storage:
+        description:
+          - V(storage) is the storage identifier of the storage to use for the C(rootfs).
+          - Mutually exclusive with V(host_path).
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:64]: Never use `V(...)` for options, only use it for values.
```suggestion
          - Mutually exclusive with O(disk_volume.host_path).
```
Same in other places.

File: plugins/modules/proxmox.py [hunk 3]
```diff
@@ -51,6 +52,41 @@
...
+        type: int
+      host_path:
+        description:
+          - V(host_path) defines a bind or device path on the PVE host to use for the C(rootfs).
+          - Mutually exclusive with V(storage), V(volume), and V(size).
+        type: path
+      options:
+        description:
+          - V(options) is a list of extra options.
+          - The options are given as key-value pairs such as C([ro=0, acl=1]).
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:87]: Shouldn't this be a dictionary then? Or is the order important, or is it possible to have keys without values?

File: plugins/modules/proxmox.py [hunk 4]
```diff
@@ -589,137 +844,184 @@ def create_instance(self, vmid, node, disk, storage, cpus, memory, swap, timeout
...
+                del kwargs["mounts"]
+            if "pubkey" in kwargs:
+                if self.version() >= LooseVersion("4.2"):
+                    kwargs["ssh-public-keys"] = kwargs["pubkey"]
+                del kwargs["pubkey"]
         else:
-            kwargs['cpus'] = cpus
-            kwargs['disk'] = disk
+            kwargs["cpus"] = cpus
+            kwargs["disk"] = disk
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:869]: There is **a lot** of unrelated reformatting in this PR. Can you please remove that? It makes reviewing it unnecessary hard.

File: plugins/modules/proxmox.py [hunk 5]
```diff
@@ -51,6 +52,41 @@
       - See U(https://pve.proxmox.com/wiki/Linux_Container) for a full description.
       - Should not be used in conjunction with O(storage).
     type: str
+  disk_volume:
+    description:
+      - Specify a hash/dictionary of the C(rootfs) disk.
+      - See U(https://pve.proxmox.com/wiki/Linux_Container#pct_mount_points) for a full description.
+    type: dict
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:59]:
```suggestion
    type: dict
    version_added: 9.2.0
```
Same for the other new option.

File: plugins/modules/proxmox.py [hunk 6]
```diff
@@ -728,115 +1030,204 @@ def umount_instance(self, vm, vmid, timeout):
...
+        required_together=[("api_token_id", "api_token_secret")],
+        required_one_of=[("api_password", "api_token_id")],
+        mutually_exclusive=[
+            (
+                "clone",
+                "ostemplate",
+                "update",
+            ),  # Creating a new container is done either by cloning an existing one, or based on a template.
+            ("disk", "disk_volume", "storage"),
+            ("mounts", "mount_volumes"),
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:1157]: If some options are mutually exclusive, this must also be documented in `DOCUMENTATION` for these options.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: plugins/modules/proxmox.py (hunk 3 above)
Comment (Lithimlin) [reply plugins/modules/proxmox.py:87]: This could be a dictionary, correct. Though then it would be slightly more work converting it to the correct string. I'll see if I can change it quickly.

File: plugins/modules/proxmox.py (hunk 4 above)
Comment (Lithimlin) [reply plugins/modules/proxmox.py]: This *should* be fixed now. Sorry about that. VS Code auto-formats files for me, but luckily I had it as a separate commit.

File: plugins/modules/proxmox.py (hunk 3 above)
Comment (Lithimlin) [reply plugins/modules/proxmox.py]: The hardest part for this was getting the documentation right.

File: plugins/modules/proxmox.py [hunk 7]
```diff
@@ -49,8 +49,44 @@
...
+        type: int
+      host_path:
+        description:
+          - O(disk_volume.host_path) defines a bind or device path on the PVE host to use for the C(rootfs).
+          - Mutually exclusive with O(disk_volume.storage), O(disk_volume.volume), and O(disk_volume.size).
+        type: path
+      options:
+        description:
+          - O(disk_volume.options) is a dict of extra options.
+          - The value of any given option must be a string, e.g. V("1").
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:88]:
```suggestion
          - The value of any given option must be a string, for example V("1").
```

File: plugins/modules/proxmox.py [hunk 8]
```diff
@@ -89,8 +125,56 @@
...
+        type: path
+      mountpoint:
+        description:
+          - O(mount_volumes[].mountpoint) is the mount point of the volume.
+        type: path
+        required: true
+      options:
+        description:
+          - O(mount_volumes[].options) is a dict of extra options.
+          - The value of any given option must be a string, e.g. V("1").
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:176]:
```suggestion
          - The value of any given option must be a string, for example V("1").
```

File: plugins/modules/proxmox.py [hunk 9]
```diff
@@ -518,22 +756,29 @@ def update_config(self, vmid, node, disk, cpus, memory, swap, **kwargs):
...
-        kwargs = dict((k, v) for k, v in kwargs.items() if v is not None)
+        kwargs = dict((key, val) for key, val in kwargs.items() if val is not None)
 
         if cpus is not None:
             kwargs["cpulimit"] = cpus
         if disk is not None:
-            kwargs["rootfs"] = disk
+            kwargs["disk_volume"] = parse_disk_string(disk)
+        if "disk_volume" in kwargs:
+            disk_dict = build_volume(key="rootfs", **kwargs.pop("disk_volume"))
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:766]: Here you should make sure that the values in `disk_volume` are all strings.

It's probably best to convert them to a string (and maybe emit a warning if they aren't).

File: plugins/modules/proxmox.py [hunk 10]
```diff
@@ -518,22 +756,29 @@ def update_config(self, vmid, node, disk, cpus, memory, swap, **kwargs):
...
-            del kwargs["netif"]
+            kwargs.update(kwargs.pop("netif"))
         if "mounts" in kwargs:
-            kwargs.update(kwargs["mounts"])
-            del kwargs["mounts"]
+            kwargs["mount_volumes"] = convert_mounts(kwargs.pop("mounts"))
+        if "mount_volumes" in kwargs:
+            mounts_list = kwargs.pop("mount_volumes")
+            for mount_config in mounts_list:
+                key = mount_config.pop("id")
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:779]: Same here for `mount_config`.

File: plugins/modules/proxmox.py [hunk 11]
```diff
@@ -518,22 +756,35 @@ def update_config(self, vmid, node, disk, cpus, memory, swap, **kwargs):
...
         if "mounts" in kwargs:
-            kwargs.update(kwargs["mounts"])
-            del kwargs["mounts"]
+            kwargs["mount_volumes"] = convert_mounts(kwargs.pop("mounts"))
+        if "mount_volumes" in kwargs:
+            mounts_list = kwargs.pop("mount_volumes")
+            for mount_config in mounts_list:
+                if not all(isinstance(val, str) for val in mount_config.values()):
+                    self.module.warn("All mount_volumes values must be strings. Converting non-string values to strings.")
+                    mount_config = {key: str(val) for key, val in mount_config.items()}
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:784]: Better use `string_types` instead of `str` (`from ansible.module_utils.six import string_types`), and use `to_text()` instead of `str()` (`from ansible.module_utils.common.text.converters import to_text`).
```suggestion
                if not all(isinstance(val, string_types) for val in mount_config.values()):
                    self.module.warn("All mount_volumes values must be strings. Converting non-string values to strings.")
                    mount_config = {key: to_text(val) for key, val in mount_config.items()}
```
Same further below.

File: changelogs/fragments/8542-fix-proxmox-volume-handling.yml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - proxmox - fix idempotency on creation of mount volumes using Proxmox' special ``<storage>:<size>`` syntax (https://github.com/ansible-collections/community.general/issues/8407, https://github.com/ansible-collections/community.general/pull/8542).
+minor_changes:
+  - proxmox - add ``disk_volume`` and ``mount_volumes`` keys for better readability (https://github.com/ansible-collections/community.general/pull/8542).
+  - proxmox - translate the old ``disk`` and ``mounts`` keys to the new handling internally (https://github.com/ansible-collections/community.general/pull/8542).
```
Comment (russoz) [inline changelogs/fragments/8542-fix-proxmox-volume-handling.yml:5]: I believe you should break this PR into two parts: one for the bugfix, the other for the new features. The bugfix gets to be backported to previous versions, while features do not. 
@felixfontein please confirm

File: plugins/modules/proxmox.py [hunk 13]
```diff
@@ -89,8 +125,56 @@
...
+      - Specify additional mounts (separate disks) for the container. As a hash/dictionary defining mount points.
+      - See U(https://pve.proxmox.com/wiki/Linux_Container#pct_mount_points) for a full description.
+      - This Option is mutually exclusive with O(mounts).
+    type: list
+    elements: dict
+    version_added: 9.2.0
+    suboptions:
+      id:
+        description:
+          - O(mount_volumes[].id) is the identifier of the mount point written as C(mp[n])
```
Comment (russoz) [inline plugins/modules/proxmox.py:142]:
```suggestion
          - O(mount_volumes[].id) is the identifier of the mount point written as C(mp[n]).
```

File: plugins/modules/proxmox.py [hunk 14]
```diff
@@ -101,8 +185,8 @@
     type: bool
   storage:
     description:
-      - target storage
-      - Should not be used in conjunction with O(disk).
+      - Target storage
```
Comment (russoz) [inline plugins/modules/proxmox.py:188]:
```suggestion
      - Target storage.
```

File: plugins/modules/proxmox.py [hunk 15]
```diff
@@ -501,6 +615,130 @@ def update_config(self, vmid, node, disk, cpus, memory, swap, **kwargs):
...
+
+            # 2. Build valid string
+            if mountpoint is not None:
+                vol_string += ",mp={mountpoint}".format(mountpoint=mountpoint)
+            if options is not None:
+                vol_string += ","
+                vol_string += ",".join(map("=".join, options.items()))
+            if kwargs:
+                vol_string += ","
+                vol_string += ",".join(map("=".join, kwargs.items()))
```
Comment (russoz) [inline plugins/modules/proxmox.py:738]: Another way to express that (feel free to ignore if you don't like it) would be:
```suggestion
            vol_string = ",".join(
                [] if host_path is None else [host_path] +
                [] if mountpoint is None else ["mp={0}".format(mountpoint)] +
                [] if options is None else [map("=".join, options.items())] +
                [] if not kwargs else [map("=".join, kwargs.items())]
            )
```
It is more concise, but not necessarily better.

**Review by russoz — COMMENTED** [review]
hi @Lithimlin 
Thanks for your contribution! I got a few comments on it, see below.

File: changelogs/fragments/8542-fix-proxmox-volume-handling.yml (hunk 12 above)
Comment (Lithimlin) [reply changelogs/fragments/8542-fix-proxmox-volume-handling.yml:5]: Currently, those changes are intertwined as I first convert the old syntax to the new one and then do the idempotency checks. Breaking this PR apart would consist of just removing the ability to *use* the new keywords; I'm not sure if that's worth it.

Also, some people *may* use this bug, though then it's a question of whether we want to support that behavior.

Comment (felixfontein) [reply changelogs/fragments/8542-fix-proxmox-volume-handling.yml:5]: I think it's OK to keep bugfix and feaure combined here. (In general it's better to separate them.)

> Also, some people *may* use this bug

https://xkcd.com/1172/ :)

Comment (russoz) [reply changelogs/fragments/8542-fix-proxmox-volume-handling.yml:5]: Fair enough :-)

**Review by russoz — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
If nobody objects I'll merge this tomorrow.

**Comment by felixfontein** [comment]
@Lithimlin thanks for your contribution!
@russoz thanks for reviewing!

**Comment by Lithimlin** [comment]
I noticed a problem that I hadn't tested after my last commit:
The string conversion turns all empty (`None`) entries into strings which messes up things down the line.
Additionally, I documented the  `size` parameters as integers, but with that the module will always throw a warning, saying that all variables must be strings. This happens even if the values are passed as strings because the module converts them back to integers.

@felixfontein since this was your comment, I'd like to talk about the conversion again.
I don't think it was even necessary since `build_volume()` will always convert it to a string anyway.
